cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,415 @@
+   */
+  [[nodiscard]] std::optional<device_span<int const>> depths() const
+  {
+    return d_depths.size() ? std::optional<device_span<int const>>(d_depths) : std::nullopt;
+  }
+
+  /**
+   * @brief Whether the table has any nullable column
+   *
+   */
+  [[nodiscard]] bool has_nulls() const { return _has_nulls; }
+
+ private:
+  using table_device_view_owner =
+    std::invoke_result_t<decltype(table_device_view::create), table_view, rmm::cuda_stream_view>;
+
+  std::unique_ptr<table_device_view_owner> d_t;
+  rmm::device_uvector<order> d_column_order;
+  rmm::device_uvector<null_order> d_null_precedence;
+  rmm::device_uvector<size_type> d_depths;
+  bool _has_nulls;
+};
+
+/**
+ * @brief An owning object that can be used to lexicographically compare two rows of the same table
+ *
+ * This class can take a table_view and preprocess certain columns to allow for lexicographical
+ * comparison. The preprocessed table and temporary data required for the comparison are created and
+ * owned by this class.
+ *
+ * This class can then provide a functor object that can used on the device.
+ * The object of this class must outlive the usage of the device functor.
+ */
+class self_comparator {
+ public:
+  /**
+   * @brief Construct an owning object for performing a lexicographic comparison between two rows of
+   * the same table.
+   *
+   * @param table The table to compare
+   * @param column_order Optional, host array the same length as a row that indicates the desired
+   * ascending/descending order of each column in a row. If empty, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If empty, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   * @param stream The stream to construct this object on. Not the stream that will be used for
+   * comparisons using this object.
+   */
+  self_comparator(table_view const& t,
+                  host_span<order const> column_order         = {},
+                  host_span<null_order const> null_precedence = {},
+                  rmm::cuda_stream_view stream                = rmm::cuda_stream_default)
+    : d_t{std::make_shared<preprocessed_table>(t, column_order, null_precedence, stream)}
+  {
+  }
+
+  /**
+   * @brief Construct an owning object for performing a lexicographic comparison between two rows of
+   * the same preprocessed table.
+   *
+   * @param t A table preprocessed for lexicographic comparison
+   */
+  self_comparator(std::shared_ptr<preprocessed_table> t) : d_t{std::move(t)} {}
+
+  /**
+   * @brief Ger the comparison operator to use on the device
+   *
+   * @tparam Nullate Optional, A cudf::nullate type describing how to check for nulls.
+   */
+  template <typename Nullate>
+  device_row_comparator<Nullate> device_comparator(Nullate nullate)
```

Comment:
For the `YES/NO` cases it is default constructible, but `DYNAMIC` is not. So this should remain convenient for the static cases and only require input for dynamic. 
```suggestion
  device_row_comparator<Nullate> device_comparator(Nullate nullate = {})
```